### PR TITLE
Merge claims from userinfo and ID Token correctly

### PR DESCRIPTION
**Pull request by jkakavas**
Enhance the handling of merging the claims sets of the
ID Token and the UserInfo response. JsonObject#merge would throw a
runtime exception when attempting to merge two objects with the
same key and different values. This could happen for an OP that
returns different vales for the same claim in the ID Token and the
UserInfo response ( Google does that for profile claim ).
If a claim is contained in both sets, we attempt to merge the
values if they are objects or arrays, otherwise the ID Token claim
value takes precedence and overwrites the userinfo response.

### CODE REVIEW

**Comment by jkakavas** [comment]
@elasticmachine run elasticsearch-ci/1

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java [hunk 1]
```diff
@@ -606,6 +621,74 @@ private void setMetadataFileWatcher(String jwkSetPath) throws IOException {
...
+    private static JSONObject merge(JSONObject jsonObject1, Object jsonObject2) {
+        if (jsonObject2 == null)
+            return jsonObject1;
+        if (jsonObject2 instanceof JSONObject)
+            return merge(jsonObject1, (JSONObject) jsonObject2);
+        throw new IllegalStateException("Error while merging ID token and userinfo claims. " +
+            "Cannot merge JSONObject with [" + jsonObject2.getClass().getName() + "]");
+    }
+
+    private static JSONArray merge(JSONArray jsonArray1, Object jsonArray2) {
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java]: Nit: I found it confusing to follow the multiple overloaded "merge" methods. Could we rename these to `mergeArrays` and `mergeObjects` or something like that?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java [hunk 2]
```diff
@@ -606,6 +621,74 @@ private void setMetadataFileWatcher(String jwkSetPath) throws IOException {
...
+        for (Map.Entry<String, Object> entry : userInfo.entrySet()) {
+            if (idToken.containsKey(entry.getKey()) == false) {
+                idToken.put(entry.getKey(), entry.getValue());
+            }
+        }
+        return idToken;
+    }
+
+    private static JSONObject merge(JSONObject jsonObject1, Object jsonObject2) {
+        if (jsonObject2 == null)
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java]: Nit: Braces `{` please.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java [hunk 3]
```diff
@@ -606,6 +621,74 @@ private void setMetadataFileWatcher(String jwkSetPath) throws IOException {
...
+            Object value1 = entry.getValue();
+            Object value2 = userInfo.get(entry.getKey());
+            if (value2 == null) {
+                continue;
+            }
+            if (value1 instanceof JSONArray) {
+                idToken.put(entry.getKey(), merge((JSONArray) value1, value2));
+            } else if (value1 instanceof JSONObject) {
+                idToken.put(entry.getKey(), merge((JSONObject) value1, value2));
+
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java]: Nit: empty newline

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java [hunk 4]
```diff
@@ -606,6 +621,74 @@ private void setMetadataFileWatcher(String jwkSetPath) throws IOException {
         watcherService.add(watcher, ResourceWatcherService.Frequency.MEDIUM);
     }
 
+    /**
+     * Merges the JsonObject with the claims of the ID Token with the JsonObject with the claims of the UserInfo response. This is
+     * necessary as some OPs return slightly different values for some claims (i.e. Google for the profile picture) and
+     * {@link JSONObject#merge(Object)} would throw a runtime exception. The merging is performed based on the following rules:
+     * <ul>
+     * <li>If the values for a given claim are strings, the value from the ID Token is retained</li>
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java]: Can a token not also have numbers and booleans?
```suggestion
     * <li>If the values for a given claim are primitives (of the the same type), the value from the ID Token is retained</li>
```

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticatorTests.java [hunk 5]
```diff
@@ -632,6 +635,85 @@ public void testImplicitFlowFailsWithUnsignedJwt() throws Exception {
...
+            .claim("name", "Jane")
+            .claim("email", "jane.doe@mail.com")
+            .subject(subject)
+            .build()
+            .toJSONObject();
+
+        OpenIdConnectAuthenticator.merge(idTokenObject, overwriteUserInfo);
+        assertThat(idTokenObject.getAsString("email"), equalTo("jane.doe@example.com"));
+        assertThat(idTokenObject.getAsString("profile"), equalTo("https://test-profiles.com/jane.doe"));
+    }
```
Comment (tvernum) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticatorTests.java]: This doesn't seem to test array or nested-object merging.

**Review by tvernum — APPROVED** [review]
LGTM, but I'd like to see additional test scenarios.

**Pull request merged**